File: fsttm/aec.py
"""
PipeWire/PulseAudio echo cancellation via module-echo-cancel.

Loads the module on enter, unloads on exit.  The two virtual devices created:
  fsttm_ec_source  — mic with TTS echo removed  → PyAudio capture
  fsttm_ec_sink    — virtual sink; TTS plays here so the AEC has the reference

Optionally chains RNNoise (LADSPA) on top of the AEC output for extra noise
suppression — enable via EchoCancelSession(rnnoise=True) / config aec.rnnoise.
"""
from __future__ import annotations  # PEP 563: lazy annotations for Python 3.8
import os
import subprocess

# Fallback master source if auto-detection finds no Jabra. Kept for back-compat;
# overridden by detect_jabra_source() / an explicit source_master.
JABRA_SOURCE = (
    "alsa_input.usb-0b0e_Jabra_Link_370_70BF928749C2-00.mono-fallback"
)
EC_SOURCE  = "fsttm_ec_source"
EC_SINK    = "fsttm_ec_sink"
RNN_SOURCE = "fsttm_rnnoise_source"   # AEC + RNNoise chain output


def detect_jabra_source(fallback: str = JABRA_SOURCE) -> str:
    """Return the first PulseAudio input source containing 'Jabra'.

    Hardware varies (Link 370 dongle, EVOLVE 20 headset, …), so match by name
    rather than hardcoding one USB id. Falls back to `fallback` if none found.
    """
    try:
        r = subprocess.run(['pactl', 'list', 'short', 'sources'],
                           capture_output=True, text=True, check=False)
        for line in r.stdout.splitlines():
            name = line.split('\t')[1] if '\t' in line else ''
            if 'Jabra' in name and name.startswith('alsa_input'):
                return name
    except Exception:
        pass
    return fallback


def _load_rnnoise_chain(source_master: str) -> str | None:
    """
    Chain RNNoise on top of source_master via module-ladspa-source (PA compat).
    Uses the system-installed librnnoise_ladspa.so. Returns the pactl module ID,
    or None if the plugin isn't available.
    """
    if not os.path.exists("/usr/lib/ladspa/librnnoise_ladspa.so"):
        return None
    r = subprocess.run(
        ["pactl", "load-module", "module-ladspa-source",
         f"source_name={RNN_SOURCE}",
         f"master={source_master}",
         "plugin=librnnoise_ladspa",
         "label=noise_suppressor_stereo"],  # stereo matches the 2ch AEC source
        capture_output=True, text=True,
    )
    mod_id = r.stdout.strip()
    return mod_id if r.returncode == 0 and mod_id else None


class EchoCancelSession:
    def __init__(self, source_master: str | None = None,
                 enabled: bool = True, rnnoise: bool = False,
                 method: str = "auto", sink_master: str | None = None):
        # None → auto-detect the connected Jabra at enable() time.
        self._source_master = source_master or detect_jabra_source()
        self._sink_master    = sink_master   # None → PA default sink (module picks)
        self._enabled        = enabled
        self._rnnoise        = rnnoise
        self._method         = (method or "auto").lower()
        self._module_id: str | None = None
        self._rnn_module_id: str | None = None
        self._prev_source: str | None = None
        self._prev_sink: str | None = None

    def _cleanup_stale(self) -> None:
        """Unload any leftover fsttm_ec / rnnoise modules from a crashed session."""
        r = subprocess.run(['pactl', 'list', 'modules', 'short'],
                           capture_output=True, text=True)
        for line in r.stdout.splitlines():
            if (f'source_name={EC_SOURCE}' in line
                    or f'sink_name={EC_SINK}' in line
                    or f'source_name={RNN_SOURCE}' in line):
                mod_id = line.split('\t')[0].strip()
                subprocess.run(['pactl', 'unload-module', mod_id], check=False)
                print(f"AEC: cleaned up stale module {mod_id}")

    def enable(self) -> None:
        if not self._enabled:
            print("AEC disabled (config aec.enabled=false)")
            return
        self._cleanup_stale()

        def _base(method):
            cmd = [
                "pactl", "load-module", "module-echo-cancel",
                f"aec_method={method}",
                f"source_name={EC_SOURCE}",
                f"sink_name={EC_SINK}",
                f"source_master={self._source_master}",
            ]
            if self._sink_master:
                cmd.append(f"sink_master={self._sink_master}")
            return cmd

        # webrtc gives AEC + built-in noise suppression; speex is echo-only but
        # loads reliably where webrtc's init fails (some Jetson PA builds). Try in
        # order of preference; "auto" walks the whole list, an explicit method
        # tries only its variants.
        WEBRTC_TUNED = _base("webrtc") + [
            "aec_args=analog_gain_control=0 digital_gain_control=1 "
            "noise_suppression=1 voice_detection=1",
        ]
        attempts = []
        if self._method in ("auto", "webrtc"):
            attempts += [("webrtc+ns", WEBRTC_TUNED), ("webrtc", _base("webrtc"))]
        if self._method in ("auto", "speex"):
            attempts += [("speex", _base("speex"))]

        last_err = ""
        for label, cmd in attempts:
            result = subprocess.run(cmd, capture_output=True, text=True)
            mod = result.stdout.strip()
            if result.returncode == 0 and mod:
                self._module_id = mod
                self._method_used = label
                print(f"AEC enabled [{label}] (module {mod}): "
                      f"{EC_SOURCE} / {EC_SINK}")
                self._maybe_chain_rnnoise()
                self._route_default()
                return
            last_err = result.stderr.strip() or f"{label} init failed"

        raise RuntimeError(
            f"AEC: module-echo-cancel failed (method={self._method}) for master "
            f"{self._source_master!r}: {last_err}")

    def _maybe_chain_rnnoise(self) -> None:
        """Chain RNNoise on top of the AEC output when requested. Sets the
        routed source so _route_default() makes it the PulseAudio default."""
        self._active_source = EC_SOURCE
        if not self._rnnoise:
            return
        rnn_id = _load_rnnoise_chain(EC_SOURCE)
        if rnn_id:
            self._rnn_module_id = rnn_id
            self._active_source = RNN_SOURCE
            import time
            time.sleep(0.3)   # let PipeWire register the node
            print(f"  RNNoise chained: {EC_SOURCE} → {RNN_SOURCE}")
        else:
            print("  RNNoise: /usr/lib/ladspa/librnnoise_ladspa.so not found, skipping")

    def _route_default(self) -> None:
        """Make the (RNNoise or EC) source and EC sink the PulseAudio defaults.

        PyAudio cannot address `fsttm_ec_source` by name (it only exposes raw
        ALSA hw devices plus the generic `pulse`/`default` devices). Routing the
        defaults to the EC source/sink lets capture via the `pulse` device pick
        up the echo-cancelled mic, and playback via `pulse` reach the EC sink.
        """
        src = getattr(self, '_active_source', EC_SOURCE)
        # Remember current defaults so we can restore them on disable().
        info = subprocess.run(['pactl', 'info'], capture_output=True, text=True)
        for line in info.stdout.splitlines():
            if line.startswith('Default Source:'):
                self._prev_source = line.split(':', 1)[1].strip()
            elif line.startswith('Default Sink:'):
                self._prev_sink = line.split(':', 1)[1].strip()
        subprocess.run(['pactl', 'set-default-source', src], check=False)
        subprocess.run(['pactl', 'set-default-sink', EC_SINK], check=False)
        print(f"AEC: default source→{src}, sink→{EC_SINK}")

    def disable(self) -> None:
        if not self._enabled:
            return
        # Restore previous defaults before unloading the modules.
        if getattr(self, '_prev_source', None):
            subprocess.run(['pactl', 'set-default-source', self._prev_source],
                           check=False)
        if getattr(self, '_prev_sink', None):
            subprocess.run(['pactl', 'set-default-sink', self._prev_sink],
                           check=False)
        if self._rnn_module_id:
            subprocess.run(["pactl", "unload-module", self._rnn_module_id],
                           check=False)
            self._rnn_module_id = None
        if self._module_id:
            subprocess.run(["pactl", "unload-module", self._module_id],
                           check=False)
            print(f"AEC disabled (module {self._module_id})")
            self._module_id = None

    def __enter__(self):
        self.enable()
        return self

    def __exit__(self, *_):
        self.disable()

File: go/internal/aec/aec.go
// Package aec — PulseAudio/PipeWire echo cancellation + noise suppression,
// a port of fsttm/aec.py.
//
// Loads module-echo-cancel at start, unloads at Close. Virtual devices:
//
//	fsttm_ec_source — mic with the TTS echo removed  → capture
//	fsttm_ec_sink   — TTS plays here so the canceller has its reference
//
// The webrtc method carries built-in NOISE SUPPRESSION (aec_args
// noise_suppression=1) — the first line of defense against the robot's
// twelve motors — and RNNoise (LADSPA) can be chained on top of the AEC
// output for the rest (config aec.rnnoise). The chain's output source and
// the EC sink are made the PulseAudio DEFAULTS, so the malgo capture and
// playback devices pick them up without any name plumbing; previous
// defaults are restored on Close.
package aec

import (
	"fmt"
	"os"
	"os/exec"
	"strings"
	"time"
)

const (
	ECSource  = "fsttm_ec_source"
	ECSink    = "fsttm_ec_sink"
	RNNSource = "fsttm_rnnoise_source" // AEC + RNNoise chain output
	ladspaSO  = "/usr/lib/ladspa/librnnoise_ladspa.so"
)

type Config struct {
	Enabled      bool
	RNNoise      bool   // chain RNNoise LADSPA on the AEC output
	Method       string // "auto" | "webrtc" | "speex"
	SourceMaster string // raw mic; "" → PulseAudio default source
}

type Session struct {
	cfg        Config
	moduleID   string
	rnnID      string
	prevSource string
	prevSink   string
	MethodUsed string
	Active     bool
}

func pactl(args ...string) (string, error) {
	out, err := exec.Command("pactl", args...).CombinedOutput()
	return strings.TrimSpace(string(out)), err
}

// Start loads the echo-cancel module (webrtc+NS → webrtc → speex), chains
// RNNoise when requested, and routes the PulseAudio defaults.
func Start(cfg Config) (*Session, error) {
	s := &Session{cfg: cfg}
	if !cfg.Enabled {
		fmt.Println("AEC disabled (config aec.enabled=false)")
		return s, nil
	}
	s.cleanupStale()

	base := func(method string) []string {
		cmd := []string{"load-module", "module-echo-cancel",
			"aec_method=" + method,
			"source_name=" + ECSource,
			"sink_name=" + ECSink}
		if cfg.SourceMaster != "" {
			cmd = append(cmd, "source_master="+cfg.SourceMaster)
		}
		return cmd
	}
	method := strings.ToLower(cfg.Method)
	if method == "" {
		method = "auto"
	}
	type attempt struct {
		label string
		args  []string
	}
	var attempts []attempt
	if method == "auto" || method == "webrtc" {
		attempts = append(attempts,
			attempt{"webrtc+ns", append(base("webrtc"),
				"aec_args=analog_gain_control=0 digital_gain_control=1 "+
					"noise_suppression=1 voice_detection=1")},
			attempt{"webrtc", base("webrtc")})
	}
	if method == "auto" || method == "speex" {
		attempts = append(attempts, attempt{"speex", base("speex")})
	}

	var lastErr string
	for _, a := range attempts {
		out, err := pactl(a.args...)
		if err == nil && out != "" {
			s.moduleID = out
			s.MethodUsed = a.label
			s.Active = true
			fmt.Printf("AEC enabled [%s] (module %s): %s / %s\n",
				a.label, out, ECSource, ECSink)
			s.maybeChainRNNoise()
			s.routeDefaults()
			return s, nil
		}
		lastErr = out
	}
	return nil, fmt.Errorf("aec: module-echo-cancel failed (method=%s): %s",
		method, lastErr)
}

func (s *Session) cleanupStale() {
	out, _ := pactl("list", "modules", "short")
	for _, line := range strings.Split(out, "\n") {
		if strings.Contains(line, "source_name="+ECSource) ||
			strings.Contains(line, "sink_name="+ECSink) ||
			strings.Contains(line, "source_name="+RNNSource) {
			id := strings.TrimSpace(strings.SplitN(line, "\t", 2)[0])
			_, _ = pactl("unload-module", id)
			fmt.Printf("AEC: cleaned up stale module %s\n", id)
		}
	}
}

func (s *Session) maybeChainRNNoise() {
	if !s.cfg.RNNoise {
		return
	}
	if _, err := os.Stat(ladspaSO); err != nil {
		fmt.Println("  RNNoise: " + ladspaSO + " not found, skipping")
		return
	}
	out, err := pactl("load-module", "module-ladspa-source",
		"source_name="+RNNSource,
		"master="+ECSource,
		"plugin=librnnoise_ladspa",
		"label=noise_suppressor_stereo") // stereo matches the 2ch AEC source
	if err == nil && out != "" {
		s.rnnID = out
		time.Sleep(300 * time.Millisecond) // let PipeWire register the node
		fmt.Printf("  RNNoise chained: %s → %s\n", ECSource, RNNSource)
	}
}

// ActiveSource is the capture chain's output — RNNoise when chained, else
// the raw EC source.
func (s *Session) ActiveSource() string {
	if s.rnnID != "" {
		return RNNSource
	}
	return ECSource
}

func (s *Session) routeDefaults() {
	// Remember the current defaults so Close can restore them.
	info, _ := pactl("info")
	for _, line := range strings.Split(info, "\n") {
		if strings.HasPrefix(line, "Default Source:") {
			s.prevSource = strings.TrimSpace(strings.SplitN(line, ":", 2)[1])
		} else if strings.HasPrefix(line, "Default Sink:") {
			s.prevSink = strings.TrimSpace(strings.SplitN(line, ":", 2)[1])
		}
	}
	_, _ = pactl("set-default-source", s.ActiveSource())
	_, _ = pactl("set-default-sink", ECSink)
	fmt.Printf("AEC: default source→%s, sink→%s\n", s.ActiveSource(), ECSink)
}

func (s *Session) Close() {
	if !s.Active {
		return
	}
	if s.prevSource != "" {
		_, _ = pactl("set-default-source", s.prevSource)
	}
	if s.prevSink != "" {
		_, _ = pactl("set-default-sink", s.prevSink)
	}
	if s.rnnID != "" {
		_, _ = pactl("unload-module", s.rnnID)
		s.rnnID = ""
	}
	if s.moduleID != "" {
		_, _ = pactl("unload-module", s.moduleID)
		fmt.Printf("AEC disabled (module %s)\n", s.moduleID)
		s.moduleID = ""
	}
	s.Active = false
}
